ROCANA-1481: Log rotation naming different from other log files

adam-rocana opened the pull request:
This commit uses a date suffix for log files and rotates the current log file when the date changes. Features like rotation on max lines and file size are preserved.

alanctgardner commented:
I like the implementation, it's nice and clean. The only problem I see is there's no way to get the original behaviour - the filename format is now fixed as `filename.<date>.<4 digit counter>`. Typically for agent changes:

- user-facing behaviour shouldn't change between minor releases (the default log file name in 1.2 and 1.3 should be the same)
- there should be an option to switch back to the old behaviour that field can use, in case we don't know the full impact of a change

We do make exceptions for minor changes. I'll defer to @MarkRoddy on this one: should we change the default, and should we expose a way to use the old pattern?

MarkRoddy commented on filelog.go:
```diff
@@ -136,28 +163,42 @@ func (w *FileLogWriter) Rotate() {
+	// If necessary, add integer suffix
+	for i := 1; i < 10000; i++ {
+		fullNameWithSuffix := fmt.Sprintf("%s.%s.%04d", filename, suffix, i)
+		if _, err := os.Stat(fullNameWithSuffix); os.IsNotExist(err) {
```
What if there's an error stat'ing the directory?  We might want to check for that condition, otherwise the output is going to be `Unable to obtain next logfile increment for %s.%s\n", filename, suffix` which isn't very helpful from the user's perspective.

adam-rocana replied:
@MarkRoddy Actually, the error is `permission denied` because the file isn't created until later. This section just attempts to discover if the file doesn't exist. I *think* if the directory can't be listed, `os.IsNotExist(err)` would return `true`, or at least that seems to be the behavior I'm seeing in my tests.

alanctgardner replied:
I would double-check about `permission denied` being `os.IsNotExist`, that seems really unexpected. I think we should clarify the `next logfile increment` error and include the error from the last call to `Stat`, that would make it clearer to users what's going wrong.

adam-rocana replied:
@alanctgardner You're right about `os.IsNotExist` - I wrote a quick test program to verify. Also, I updated the wording to be more clear and included the last `Stat` error.

MarkRoddy commented:
Unfortunately @alanctgardner's right about the backwards compatibility.  Sorry I hadn't thought of that earlier @adam-rocana.  How much effort do you think it would be to make it an option to use indexes or dates as the suffix?

adam-rocana commented:
@MarkRoddy Your directory comment made me think a little more about error cases there, so I created `makeDirectory` that creates the directory before each rotation, and attempts a `stat` to catch any obviously wrong permissions. I don't think we can portably and easily determine whether a file can be created within a directory, so in the case of a successful `stat` but no actual permission, the error will be something like

```
FileLogWriter("/path/to/file.log"): open /path/to/file.log: permission denied
```

The error will be the same whether the file permissions are bad or somewhere up the directory chain.

Also, I fixed a bug where failure in rotation would cause a bad file descriptor by moving the `Close` call to just after `OpenFile`.

@alanctgardner Thanks for pointing out compatibility! I think this is a minor change, and the risk is low, but I added an option just in case. I made the default date-based rotations because integer-based is just broken. We can choose either as the default in follow-on code that uses log4go.

MarkRoddy commented on filelog.go:
```diff
@@ -62,26 +98,26 @@ func (w *FileLogWriter) Close() {
+		filename:            fname,
+		format:              "[%D %T] [%L] (%S) %M",
+		rotate:              rotate,
+		rotateIntegerSuffix: false,
```
We try to take a very light hand on formatting with the 3rd party libraries. It makes pulling in upstream changes very difficult. The upstream for this project isn't very active so I don't think it's that big of deal, but generally speaking we should be avoiding formatting only changes in forks.

MarkRoddy commented on filelog.go:
```diff
@@ -136,28 +177,58 @@ func (w *FileLogWriter) Rotate() {
-		w.file.Close()
+// Generate the next filename for rotation using integer suffix
+func (w *FileLogWriter) nextIntegerFilename(filename string) (string, error) {
+	for i := 1; i <= 999; i++ {
```
Do we *need* an upper bound here?  Barring that, should we maybe crank it up a ton?

adam-rocana replied:
We could crank up the number, but eventually you run into a situation where that loop could churn for a long, long time. The pathological case could cause a lot of unnecessary disk IO.

MarkRoddy replied:
I guess it depends on what happens when we hit 999 files.  If it's "we stop logging" then that's a problem. If it's "we keep logging but we stop rotating" that's not really great but I think we can live with it.

MarkRoddy commented on config.go:
```diff
@@ -249,6 +250,8 @@ func xmlToFileLogWriter(filename string, props []xmlProperty, enabled bool) (*Fi
 		case "rotate":
 			rotate = strings.Trim(prop.Value, " \r\n") != "false"
+		case "integersuffix":
+			integerSuffix = strings.Trim(prop.Value, " \r\n") != "false"
```
Unfortunately we have to retain the default behavior, and not just make it available. A good test is, if user upgraded the agent and didn't change their configuration file would they still see the same behavior.

Is there a way we can clean up the integer suffix so it's less of a threat to agent health?  I know it sucks to have to clean up the implementation we actively want to get rid of, but it's unfortunately necessary for the backwards compatibility promises we make to our customers.

adam-rocana replied:
I changed the default behavior within the library, but planned to leave the old behavior as default in the agent. So there shouldn't be a backwards-incompatible change, but new uses of log4go will use the preferred behavior. In other words, if a user upgraded an agent with no other changes, they *would* see the same behavior.

One of the changes here plus a forthcoming PR will actually tolerate rotation failure, so even if an integer suffix is used and we reach the 1000th rotation, log4go will continue to write to the "current" logfile.

MarkRoddy replied:
As long as the user sees the same behavior then we are good.  How are we planning on manipulating the default behavior from within the agent?

adam-rocana replied:
I looked at configuring the agent to default to integer-based suffixes (while keeping log4go at date-based), but it looks like too much work because log4go doesn't provide an easy way to discover the property set in the XML. I'll just expose the option to use date-based suffixes in the agent and then submit a JIRA to switch the default in the next major release.

adam-rocana replied:
Also, a new commit to default to integer-based suffixes.

MarkRoddy commented:
@adam-rocana I think we're all set!
